README.md
# Avataaars plugin for Craft CMS 3.x

With Avataaars for Craft CMS you can create avatar icons by combining clothes, hair, emotions, accessories, and colors. This plugin is based on the Avataaars set designed by Pablo Stanley and the HTTP API provided by Dicebear.

## Requirements

This plugin requires Craft CMS 3.0.0-beta.23 or later.

## Installation

To install the plugin, follow these instructions.

1. Open your terminal and go to your Craft project:

        cd /path/to/project

2. Then tell Composer to load the plugin:

        composer require remcoov/avataaars

3. In the Control Panel, go to Settings → Plugins and click the “Install” button for Avataaars.

## Using Avataaars

In your twig templates, you can create a lot of different Avataaars by playing with all the [available settings](#available-settings). The Avataaar icon used for this Craft CMS plugin, for example, is made with the following settings:

```
{% set avataaar = craft.avataaars.avataaar({
    seed: 'avataaar',
    radius: 0,
    width: 200,
    height: 200,
    margin: 0,
    background: 'CCCCCC',
    style: 'circle',
    top: 'shortHair',
    hatColor: 'red',
    hairColor: 'black',
    accessories: 'prescription02',
    facialHair: 'majestic',
    facialHairColor: 'black',
    clothes: 'hoodie',
    clothesColor: 'pink',
    eyes: 'happy',
    eyebrow: 'flat',
    mouth: 'smile',
    skin: 'light',
    svg: false
}) %}

<img src="{{ avataaar }}" />
```

## Available settings

#### seed
The value of _seed_ can be anything you like - but don't use any sensitive or personal data here!

#### radius
Avatar border radius. Max: `50`

#### width
Fixed width

#### height
Fixed height

#### margin
Avatar margin in percent. Max: `25`

#### background
Only hex (3-digit, 6-digit and 8-digit) values are allowed. For example: `FF0000`

#### style
`transparent` or `circle`

#### top
`longHair`, `shortHair`, `eyepatch`, `hat`, `hijab` or `turban`

#### hatColor
`black`, `blue`, `gray`, `heather`, `pastel`, `pink`, `red` or `white`


#### hairColor
`auburn`, `black`, `blonde`, `brown`, `pastel`, `platinum`, `red` or `gray`

#### accessories
`kurt`, `prescription01`, `prescription02`, `round`, `sunglasses` or `wayfarers`

#### facialHair
`medium`, `light`, `majestic`, `fancy` or `magnum`

#### facialHairColor
`auburn`, `black`, `blonde`, `brown`, `platinum` or `red`

#### clothes
`blazer`, `sweater`, `shirt`, `hoodie` or `overall`

#### clothesColor
`black`, `blue`, `gray`, `heather`, `pastel`, `pink`, `red` or `white`

#### eyes
`close`, `cry`, `default`, `dizzy`, `roll`, `happy`, `hearts`, `side`, `squint`, `surprised`, `wink` or `winkWacky`

#### eyebrow
`angry`, `default`, `flat`, `raised`, `sad`, `unibrow` or `up`

#### mouth
`concerned`, `default`, `disbelief`, `eating`, `grimace`, `sad`, `scream`, `serious`, `smile`, `tongue`, `twinkle` or `vomit`

#### skin
`tanned`, `yellow`, `pale`, `light`, `brown`, `darkBrown` or `black`

#### svg
When you'd like to output the Avataaar as raw SVG code, you'd want to set this to `true`. Example:

```
{% set avataaar = craft.avataaars.avataaar({
    svg: true
}) %}

{{ svg(avataaar) }}
```

If you would like to add a class to the SVG:

```
{{ svg(avataaar)|attr({ class: 'avataaar-icon' }) }}
```

## Random Avataaar

You can also output a random Avataaar by simply doing the following:

`<img src="{{ craft.avataaars.avataaar }}" />`

## Form to set Craft CMS user photo

You can render a form on the front-end to let (logged in) users set a rendered Avataaar as their Craft CMS user photo. To do this, add the following code:

```
{% set user = user ?? currentUser %}
{% if user %}

    <form method="POST" id="userPhoto-form">
        {{ craft.avataaars.userPhotoForm|raw }}
        
        <input type="hidden" name="width" value="300" />
        <input type="hidden" name="height" value="300" />
        <input type="hidden" name="background" value="" /> <!-- f.e. FF0000 -->
        <input type="hidden" name="margin" value="" /> <!-- max. 25 -->

        {{ csrfInput() }}
        <div class="userPhoto-result">
            <img src='data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8Xw8AAoMBgDTD2qgAAAAASUVORK5CYII=' id="userPhoto-result-img" />
        </div>

        <input id="userPhoto-submit" type="submit" value="Set user photo" >
    </form>

{% endif %}
```

The form has a couple of hidden configurations for you as a developer to set, since these are probably too 'technical' for the user: `width`, `height`,`background` and `margin`.

When the user sets the profile picture and you want to do something on either `success` or `error`, you can use the following Javascript code:

```
window.addEventListener('userPhotoStatus', function (e) {
    console.log(e.detail.status);
}, false);
```

A note beforehand: remember to set the volume for user photo storage.

Brought to you by [remcoov](https://github.com/remcoov) and [kevinmu17](https://github.com/kevinmu17)

CHANGELOG.md
# Avataaars Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.2 - 2020-09-09
### Added
- You can now render a form to let a user set their Craft CMS user photo ([#1](https://github.com/remcoov/avataaars/issues/1))

## 1.0.1 - 2020-08-27
### Added
- You can now output the Avataaar icon as raw SVG code

## 1.0.0 - 2020-08-27
### Added
- Initial release


src/Avataaars.php
<?php
/**
 * Avataaars plugin for Craft CMS 3.x
 *
 * With Avataaars you can create avatar icons by combining clothes, hair, emotions, accessories, and colors.
 *
 * @link      https://github.com/remcoov
 * @copyright Copyright (c) 2020 remcoov
 */

namespace remcoov\avataaars;

use remcoov\avataaars\services\AvataaarsService as AvataaarsServiceService;
use remcoov\avataaars\variables\AvataaarsVariable;

use Craft;
use craft\base\Plugin;
use craft\services\Plugins;
use craft\events\PluginEvent;
use craft\console\Application as ConsoleApplication;
use craft\web\UrlManager;
use craft\services\Elements;
use craft\services\Fields;
use craft\services\Utilities;
use craft\web\twig\variables\CraftVariable;
use craft\services\Dashboard;
use craft\events\RegisterComponentTypesEvent;
use craft\events\RegisterUrlRulesEvent;

use yii\base\Event;

/**
 * Craft plugins are very much like little applications in and of themselves. We’ve made
 * it as simple as we can, but the training wheels are off. A little prior knowledge is
 * going to be required to write a plugin.
 *
 * For the purposes of the plugin docs, we’re going to assume that you know PHP and SQL,
 * as well as some semi-advanced concepts like object-oriented programming and PHP namespaces.
 *
 * https://docs.craftcms.com/v3/extend/
 *
 * @author    remcoov
 * @package   Avataaars
 * @since     1.0.0
 *
 * @property  AvataaarsServiceService $avataaarsService
 * @property  Settings $settings
 * @method    Settings getSettings()
 */
class Avataaars extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * Static property that is an instance of this plugin class so that it can be accessed via
     * Avataaars::$plugin
     *
     * @var Avataaars
     */
    public static $plugin;

    // Public Properties
    // =========================================================================

    /**
     * To execute your plugin’s migrations, you’ll need to increase its schema version.
     *
     * @var string
     */
    public $schemaVersion = '1.0.0';
    public $hasCpSettings = false;
    public $hasCpSection = false;

    // Public Methods
    // =========================================================================

    public function init()
    {
        parent::init();
        self::$plugin = $this;

        // Register our variables
        Event::on(
            CraftVariable::class,
            CraftVariable::EVENT_INIT,
            function (Event $event) {
                /** @var CraftVariable $variable */
                $variable = $event->sender;
                $variable->set('avataaars', AvataaarsVariable::class);
            }
        );

        Craft::info(
            Craft::t(
                'avataaars',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
    }

    // Protected Methods
    // =========================================================================

    /**
     * Creates and returns the model used to store the plugin’s settings.
     *
     * @return \craft\base\Model|null
     */
    protected function createSettingsModel()
    {
        return new Settings();
    }

    /**
     * Returns the rendered settings HTML, which will be inserted into the content
     * block on the settings page.
     *
     * @return string The rendered settings HTML
     */
    protected function settingsHtml(): string
    {
        return Craft::$app->view->renderTemplate(
            'avataaars/settings',
            [
                'settings' => $this->getSettings()
            ]
        );
    }
}


src/variables/AvataaarsVariable.php
<?php
/**
 * Avataaars plugin for Craft CMS 3.x
 *
 * With Avataaars you can create avatar icons by combining clothes, hair, emotions, accessories, and colors.
 *
 * @link      https://github.com/remcoov
 * @copyright Copyright (c) 2020 remcoov
 */

namespace remcoov\avataaars\variables;

use remcoov\avataaars\Avataaars;

use Craft;

/**
 * Avataaars Variable
 *
 * Craft allows plugins to provide their own template variables, accessible from
 * the {{ craft }} global variable (e.g. {{ craft.avataaars }}).
 *
 * https://craftcms.com/docs/plugins/variables
 *
 * @author    remcoov
 * @package   Avataaars
 * @since     1.0.0
 */
class AvataaarsVariable
{
    public function avataaar(array $options = [] ) : string {
        return Avataaars::$plugin->avataaarsService->avataaar($options);
    }

    public function userPhotoForm(array $options = [] ) : string {
        return Avataaars::$plugin->avataaarsService->userPhotoForm($options);
    }
}


src/controllers/UserPhotoController.php
<?php
/**
 * Avataaars plugin for Craft CMS 3.x
 *
 * With Avataaars you can create avatar icons by combining clothes, hair, emotions, accessories, and colors.
 *
 * @link      https://github.com/remcoov
 * @copyright Copyright (c) 2020 remcoov
 */

namespace remcoov\avataaars\controllers;

use remcoov\avataaars\Avataaars;

use Craft;
use craft\web\Controller;

/**
 * Avataaars Controller
 *
 * Generally speaking, controllers are the middlemen between the front end of
 * the CP/website and your plugin’s services. They contain action methods which
 * handle individual tasks.
 *
 * A common pattern used throughout Craft involves a controller action gathering
 * post data, saving it on a model, passing the model off to a service, and then
 * responding to the request appropriately depending on the service method’s response.
 *
 * Action methods begin with the prefix “action”, followed by a description of what
 * the method does (for example, actionSaveIngredient()).
 *
 * https://craftcms.com/docs/plugins/controllers
 *
 * @author    remcoov
 * @package   Avataaars
 * @since     1.0.0
 */
class UserPhotoController extends Controller
{

    // Protected Properties
    // =========================================================================

    /**
     * @var    bool|array Allows anonymous access to this controller's actions.
     *         The actions must be in 'kebab-case'
     * @access protected
     */
    protected $allowAnonymous = [];

    // Public Methods
    // =========================================================================

    public function actionGetImagePreview()
    {
        $options = $_POST;
        unset($options['CRAFT_CSRF_TOKEN']);

        return Avataaars::$plugin->avataaarsService->avataaar($options);
    }

    public function actionSetUserPhoto()
    {
        $options = $_POST;
        unset($options['CRAFT_CSRF_TOKEN']);

        $svgCode = Avataaars::$plugin->avataaarsService->avataaar($options);
        return Avataaars::$plugin->avataaarsService->setUserPhoto($svgCode);
    }

}


src/assetbundles/avataaars/AvataaarsAsset.php
<?php
/**
 * Avataaars plugin for Craft CMS 3.x
 *
 * With Avataaars you can create avatar icons by combining clothes, hair, emotions, accessories, and colors.
 *
 * @link      https://github.com/remcoov
 * @copyright Copyright (c) 2020 remcoov
 */

namespace remcoov\avataaars\assetbundles\avataaars;

use Craft;
use craft\web\AssetBundle;

/**
 * AvataaarsAsset AssetBundle
 *
 * AssetBundle represents a collection of asset files, such as CSS, JS, images.
 *
 * Each asset bundle has a unique name that globally identifies it among all asset bundles used in an application.
 * The name is the [fully qualified class name](http://php.net/manual/en/language.namespaces.rules.php)
 * of the class representing it.
 *
 * An asset bundle can depend on other asset bundles. When registering an asset bundle
 * with a view, all its dependent asset bundles will be automatically registered.
 *
 * http://www.yiiframework.com/doc-2.0/guide-structure-assets.html
 *
 * @author    remcoov
 * @package   Avataaars
 * @since     1.0.0
 */
class AvataaarsAsset extends AssetBundle
{
    // Public Methods
    // =========================================================================

    /**
     * Initializes the bundle.
     */
    public function init()
    {
        // define the path that your publishable resources live
        $this->sourcePath = "@remcoov/avataaars/assetbundles/avataaars/dist";

        // define the relative path to CSS/JS files that should be registered with the page
        // when this asset bundle is registered
        $this->js = [
            'js/Avataaars.js',
        ];

        $this->css = [
            'css/Avataaars.css',
        ];

        parent::init();
    }
}


src/services/AvataaarsService.php
<?php
/**
 * Avataaars plugin for Craft CMS 3.x
 *
 * With Avataaars you can create avatar icons by combining clothes, hair, emotions, accessories, and colors.
 *
 * @link      https://github.com/remcoov
 * @copyright Copyright (c) 2020 remcoov
 */

namespace remcoov\avataaars\services;

use remcoov\avataaars\Avataaars;

use Craft;
use craft\base\Component;
use craft\web\View; // FOR SETTING USERPIC
use craft\elements\Asset; // FOR SETTING USERPIC
use craft\helpers\Assets; // FOR SETTING USERPIC
use craft\errors\VolumeException; // FOR SETTING USERPIC

/**
 * AvataaarsService Service
 *
 * https://craftcms.com/docs/plugins/services
 *
 * @author    remcoov
 * @package   Avataaars
 * @since     1.0.0
 */
class AvataaarsService extends Component
{
    // Public Methods
    // =========================================================================

    public function avataaar(array $options = []) : string {
        
        $defaultOptions = [
            'seed' => substr(md5(rand()),0,8),
            'radius' => 0, // max 50
            'width' => NULL,
            'height' => NULL,
            'margin' => 0, // max 25
            'background' => NULL,
            'style' => 'transparent', // transparent, circle
            'top' => NULL, // longHair, shortHair, eyepatch, hat, hijab, turban
            'hatColor' => NULL, // black, blue, gray, heather, pastel, pink, red, white
            'hairColor' => NULL, // auburn, black, blonde, brown, pastel, platinum, red, gray
            'accessories' => NULL, // kurt, prescription01, prescription02, round, sunglasses, wayfarers
            'facialHair' => NULL, // medium, light, majestic, fancy, magnum
            'facialHairColor' => NULL, // auburn, black, blonde, brown, platinum, red
            'clothes' => NULL, // blazer, sweater, shirt, hoodie, overall
            'clothesColor' => NULL, // black, blue, gray, heather, pastel, pink, red, white
            'eyes' => NULL, // close, cry, default, dizzy, roll, happy, hearts, side, squint, surprised, wink, winkWacky
            'eyebrow' => NULL, // angry, default, flat, raised, sad, unibrow, up
            'mouth' => NULL, // concerned, default, disbelief, eating, grimace, sad, scream, serious, smile, tongue, twinkle, vomit
            'skin' => NULL, // tanned, yellow, pale, light, brown, darkBrown, black
            'svg' => FALSE // TRUE OR FALSE
        ];

        $finalOptions = array_merge($defaultOptions, $options);

        $url = rtrim('https://avatars.dicebear.com/api/avataaars/'.$finalOptions['seed'].'.svg?'
            .( $finalOptions['radius'] ? 'radius='.$finalOptions['radius'].'&' : NULL )
            .( $finalOptions['width'] ? 'width='.$finalOptions['width'].'&' : NULL )
            .( $finalOptions['height'] ? 'height='.$finalOptions['height'].'&' : NULL )
            .( $finalOptions['margin'] ? 'margin='.$finalOptions['margin'].'&' : NULL )
            .( $finalOptions['background'] ? 'background=%23'.$finalOptions['background'].'&' : NULL )
            .( $finalOptions['style'] ? 'style='.$finalOptions['style'].'&' : NULL )
            .( $finalOptions['top'] ? 'top[]='.$finalOptions['top'].'&' : NULL )
            .( $finalOptions['hatColor'] ? 'hatColor[]='.$finalOptions['hatColor'].'&' : NULL )
            .( $finalOptions['hairColor'] ? 'hairColor[]='.$finalOptions['hairColor'].'&' : NULL )
            .( $finalOptions['accessories'] ? 'accessories[]='.$finalOptions['accessories'].'&accessoriesChance=100&' : NULL )
            .( $finalOptions['facialHair'] ? 'facialHair[]='.$finalOptions['facialHair'].'&facialHairChance=100&' : NULL )
            .( $finalOptions['facialHairColor'] ? 'facialHairColor[]='.$finalOptions['facialHairColor'].'&' : NULL )
            .( $finalOptions['clothes'] ? 'clothes[]='.$finalOptions['clothes'].'&' : NULL )
            .( $finalOptions['clothesColor'] ? 'clothesColor[]='.$finalOptions['clothesColor'].'&' : NULL )
            .( $finalOptions['eyes'] ? 'eyes[]='.$finalOptions['eyes'].'&' : NULL )
            .( $finalOptions['eyebrow'] ? 'eyebrow[]='.$finalOptions['eyebrow'].'&' : NULL )
            .( $finalOptions['mouth'] ? 'mouth[]='.$finalOptions['mouth'].'&' : NULL )
            .( $finalOptions['skin'] ? 'skin[]='.$finalOptions['skin'].'&' : NULL )
        ,'&');

        if( $finalOptions['svg'] ) {
            $curl = curl_init();
            curl_setopt_array($curl, array(
              CURLOPT_URL => $url,
              CURLOPT_RETURNTRANSFER => true,
              CURLOPT_ENCODING => "",
              CURLOPT_MAXREDIRS => 10,
              CURLOPT_TIMEOUT => 0,
              CURLOPT_FOLLOWLOCATION => true,
              CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
              CURLOPT_CUSTOMREQUEST => "GET",
            ));
            $svg = curl_exec($curl);
            curl_close($curl);

            return $svg;
        } else {
            return $url;
        }

    }

    public function userPhotoForm(array $options = []) : string {

        $volumes = Craft::$app->getVolumes();
        $volumeUid = Craft::$app->getProjectConfig()->get('users.photoVolumeUid');

        if (!$volumeUid || ($volume = $volumes->getVolumeByUid($volumeUid)) === null) {
            throw new VolumeException(Craft::t('app',
                'The volume set for user photo storage is not valid.'));
        }
        
        $oldMode = Craft::$app->view->getTemplateMode();
        Craft::$app->view->setTemplateMode(View::TEMPLATE_MODE_CP);
        $showUserPhotoForm = Craft::$app->view->renderTemplate('avataaars/userPhotoForm');
        Craft::$app->view->setTemplateMode($oldMode);
        return $showUserPhotoForm;
        
    }

    public function setUserPhoto(string $svgCode) : bool {

        $users = Craft::$app->getUsers();
        $userId = Craft::$app->getUser()->id;
        $user = Craft::$app->users->getUserById($userId);

        $filename = Assets::prepareAssetName('avataaar.svg');
        $tempPath = Assets::tempFilePath($filename);
        file_put_contents($tempPath, $svgCode);
        $users->saveUserPhoto($tempPath, $user, $filename);

        return true;
    }

}
